File: M3Tools/Controls/People/Customers/CustomersDataGrid.vb
Imports SPPBC.M3Tools.Events
Imports SPPBC.M3Tools.Events.Customers
Imports System.ComponentModel
Imports System.Windows.Forms

' TODO: Remove old dgc for remove and edit if customs continue to work out. keep until proven
Public Class CustomersDataGrid
	Inherits DataGridView
	Public Event EditCustomer As CustomerEventHandler
	Public Event RemoveCustomer As CustomerEventHandler
	Public Event RefreshDisplay()

	'<Browsable(False)>
	'Public ReadOnly Property SelectedRowsCount As Integer
	'	Get
	'		Return SelectedRows.Count
	'	End Get
	'End Property

	<Browsable(False)>
	Public ReadOnly Property Customers As IList
		Get
			Return Rows
		End Get
	End Property

	'<DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
	'<DefaultValue(GetType(BindingSource))>
	<RefreshProperties(RefreshProperties.Repaint)>
	<AttributeProvider(GetType(IListSource))>
	<Description("Data Source to use for data grid.")>
	Public Shadows Property DataSource As BindingSource
		Get
			Return bsCustomers
		End Get
		Set(value As BindingSource)
			AutoGenerateColumns = False
			bsCustomers = CType(value, Types.CustomersBindingSource)
		End Set
	End Property

	<Browsable(False)>
	Public Shadows ReadOnly Property SelectedRows As IList
		Get
			If CustomersSelectable Then
				If chk_SelectAll.Checked Then
					Return Rows
				End If

				For Each row As DataGridViewRow In Rows
					row.Selected = CBool(row.Cells(dgc_Selection.DisplayIndex).Value)
				Next
			End If



			Return MyBase.SelectedRows
		End Get
	End Property

	'<DefaultValue("")>
	'Property Filter As String
	'	Get
	'		Return DataSource?.Filter
	'	End Get
	'	Set(value As String)
	'		DataSource.Filter = value
	'	End Set
	'End Property

	<DefaultValue(True)>
	Property AllowEditting As Boolean
		Get
			Return dgc_Edit.Visible
		End Get
		Set(value As Boolean)
			dgc_Edit.Visible = value
		End Set
	End Property

	<DefaultValue(False)>
	Property AllowAdding As Boolean
		Get
			Return AllowUserToAddRows
		End Get
		Set(value As Boolean)
			AllowUserToAddRows = value
		End Set
	End Property

	<DefaultValue(True)>
	Property AllowDeleting As Boolean
		Get
			Return dgc_Remove.Visible
		End Get
		Set(value As Boolean)
			dgc_Remove.Visible = value
		End Set
	End Property

	<DefaultValue(False)>
	Property AllowColumnReordering As Boolean
		Get
			Return AllowUserToOrderColumns
		End Get
		Set(value As Boolean)
			AllowUserToOrderColumns = value
		End Set
	End Property

	<DefaultValue(True)>
	Property CustomersSelectable As Boolean
		Get
			Return dgc_Selection.Visible
		End Get
		Set(value As Boolean)
			dgc_Selection.Visible = value
			chk_SelectAll.Visible = value
		End Set
	End Property

	Private Sub CellClicked(sender As Object, e As DataGridViewCellEventArgs) Handles Me.CellContentClick
		If e.ColumnIndex <> dgc_Edit.Index AndAlso e.ColumnIndex <> dgc_Remove.Index Then
			Return
		End If

		Dim row = Rows(e.RowIndex)
		Dim customer = CType(row.DataBoundItem, Types.Customer)

		Select Case e.ColumnIndex
			Case dgc_Edit.Index
				RaiseEvent EditCustomer(Me, New CustomerEventArgs(customer, EventType.Updated))
			Case dgc_Remove.DisplayIndex
				DeleteCustomer(Me, New DataGridViewRowCancelEventArgs(row))
		End Select
	End Sub

	Private Sub DeleteCustomer(sender As Object, e As DataGridViewRowCancelEventArgs) Handles Me.UserDeletingRow
		Dim customer = CType(e.Row.DataBoundItem, Types.Customer)

		RaiseEvent RemoveCustomer(Me, New CustomerEventArgs(customer, EventType.Deleted))
	End Sub

	Public Sub RemoveSelectedRows() Handles cms_Tools.RemoveRows
		If SelectedRows.Count < 1 Then
			Return
		End If

		Dim failed As Integer = 0
		Dim total As Integer = SelectedRows.Count

		For Each row As DataGridViewRow In SelectedRows
			Try
				DeleteCustomer(Me, New DataGridViewRowCancelEventArgs(row))
			Catch ex As Exception
				Console.WriteLine(ex.Message)
				failed += 1
				Continue For
			End Try
		Next

		If failed > 0 Then
			MessageBox.Show($"Failed to remove {failed} customer{If(failed > 1, "s", "")}", "Failed Removals", MessageBoxButtons.OK, MessageBoxIcon.Error)
		End If

		If total - failed > 0 Then
			MessageBox.Show($"Successfully removed {total - failed} customer{If(total - failed > 1, "s", "")}", "Successful Removals", MessageBoxButtons.OK, MessageBoxIcon.Information)
		End If
	End Sub

	Private Sub SelectAllCustomers(sender As Object, e As EventArgs) Handles chk_SelectAll.CheckedChanged
		For Each row As DataGridViewRow In Rows
			row.Cells(dgc_Selection.DisplayIndex).Value = chk_SelectAll.Checked
		Next
		Invalidate()
	End Sub

	Private Sub ToolsOpened(sender As Object, e As EventArgs) Handles cms_Tools.Opened
		cms_Tools.ToggleRemove(SelectedRows.Count > 0)
		cms_Tools.ToggleEdit(SelectedRows.Count > 0)
	End Sub

	Private Sub EditPerson() Handles cms_Tools.EditSelected
		If SelectedRows.Count < 1 Then
			Return
		End If

		For Each row As DataGridViewRow In SelectedRows
			CellClicked(Me, New DataGridViewCellEventArgs(dgc_Edit.Index, row.Index))
		Next
	End Sub

	Private Sub RefreshView() Handles cms_Tools.RefreshView
		Refresh()
	End Sub
End Class
'Public Class CustomersDataGrid
'	Inherits DataGridView
'	Public Event EditCustomer As CustomerEventHandler
'	Public Event RemoveCustomer As CustomerEventHandler
'	Public Event RefreshDisplay()

'	<Browsable(False)>
'	Public ReadOnly Property SelectedRowsCount As Integer
'		Get
'			Return SelectedCustomers.Count
'		End Get
'	End Property

'	<Browsable(False)>
'	Public ReadOnly Property Customers As IList
'		Get
'			Return Rows
'		End Get
'	End Property

'	'<DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
'	'<DefaultValue(GetType(BindingSource))>
'	<RefreshProperties(RefreshProperties.Repaint)>
'	<AttributeProvider(GetType(IListSource))>
'	<Description("Data Source to use for data grid.")>
'	Public Shadows Property DataSource As BindingSource
'		Get
'			Return CType(dgv_Customers.DataSource, BindingSource)
'		End Get
'		Set(value As BindingSource)
'			Me.AutoGenerateColumns = False
'			MyBase.DataSource = value
'			'dgv_Customers.Refresh()
'		End Set
'	End Property

'	<Browsable(False)>
'	Public ReadOnly Property SelectedCustomers As IList
'		Get
'			If CustomersSelectable Then
'				If chk_SelectAll.Checked Then
'					Return dgv_Customers.Rows
'				End If

'				For Each row As DataGridViewRow In dgv_Customers.Rows
'					row.Selected = CBool(row.Cells(dgc_Selection.DisplayIndex).Value)
'				Next
'			End If

'			Return dgv_Customers.SelectedRows
'		End Get
'	End Property

'	'<DefaultValue("")>
'	'Property Filter As String
'	'	Get
'	'		Return DataSource?.Filter
'	'	End Get
'	'	Set(value As String)
'	'		DataSource.Filter = value
'	'	End Set
'	'End Property

'	<DefaultValue(True)>
'	Property AllowEditting As Boolean
'		Get
'			Return dgc_Edit.Visible
'		End Get
'		Set(value As Boolean)
'			dgc_Edit.Visible = value
'		End Set
'	End Property

'	<DefaultValue(False)>
'	Property AllowAdding As Boolean
'		Get
'			Return dgv_Customers.AllowUserToAddRows
'		End Get
'		Set(value As Boolean)
'			dgv_Customers.AllowUserToAddRows = value
'		End Set
'	End Property

'	<DefaultValue(True)>
'	Property AllowDeleting As Boolean
'		Get
'			Return dgc_Remove.Visible
'		End Get
'		Set(value As Boolean)
'			dgc_Remove.Visible = value
'		End Set
'	End Property

'	<DefaultValue(False)>
'	Property AllowColumnReordering As Boolean
'		Get
'			Return dgv_Customers.AllowUserToOrderColumns
'		End Get
'		Set(value As Boolean)
'			dgv_Customers.AllowUserToOrderColumns = value
'		End Set
'	End Property

'	<DefaultValue(True)>
'	Property CustomersSelectable As Boolean
'		Get
'			Return dgc_Selection.Visible
'		End Get
'		Set(value As Boolean)
'			dgc_Selection.Visible = value
'			chk_SelectAll.Visible = value
'		End Set
'	End Property

'	Private Sub CellClicked(sender As Object, e As DataGridViewCellEventArgs) Handles dgv_Customers.CellContentClick
'		If e.ColumnIndex <> dgc_Edit.Index AndAlso e.ColumnIndex <> dgc_Remove.Index Then
'			Return
'		End If

'		Dim row = dgv_Customers.Rows(e.RowIndex)
'		Dim customer = CType(row.DataBoundItem, Types.Customer)

'		Select Case e.ColumnIndex
'			Case dgc_Edit.Index
'				RaiseEvent EditCustomer(Me, New CustomerEventArgs(customer, EventType.Updated))
'			Case dgc_Remove.DisplayIndex
'				DeleteCustomer(Me, New DataGridViewRowCancelEventArgs(row))
'		End Select
'	End Sub

'	Private Sub DeleteCustomer(sender As Object, e As DataGridViewRowCancelEventArgs) Handles dgv_Customers.UserDeletingRow
'		Dim customer = CType(e.Row.DataBoundItem, Types.Customer)

'		RaiseEvent RemoveCustomer(Me, New CustomerEventArgs(customer, EventType.Deleted))
'	End Sub

'	Public Sub RemoveSelectedRows() Handles cms_Tools.RemoveRows
'		If SelectedRowsCount < 1 Then
'			Return
'		End If

'		Dim failed As Integer = 0
'		Dim total As Integer = dgv_Customers.SelectedRows.Count

'		For Each row As DataGridViewRow In dgv_Customers.SelectedRows
'			Try
'				DeleteCustomer(Me, New DataGridViewRowCancelEventArgs(row))
'			Catch ex As Exception
'				Console.WriteLine(ex.Message)
'				failed += 1
'				Continue For
'			End Try
'		Next

'		If failed > 0 Then
'			MessageBox.Show($"Failed to remove {failed} customer{If(failed > 1, "s", "")}", "Failed Removals", MessageBoxButtons.OK, MessageBoxIcon.Error)
'		End If

'		If total - failed > 0 Then
'			MessageBox.Show($"Successfully removed {total - failed} customer{If(total - failed > 1, "s", "")}", "Successful Removals", MessageBoxButtons.OK, MessageBoxIcon.Information)
'		End If
'	End Sub

'	Private Sub SelectAllCustomers(sender As Object, e As EventArgs) Handles chk_SelectAll.CheckedChanged
'		For Each row As DataGridViewRow In dgv_Customers.Rows
'			row.Cells(dgc_Selection.DisplayIndex).Value = chk_SelectAll.Checked
'		Next
'		dgv_Customers.Invalidate()
'	End Sub

'	Private Sub ToolsOpened(sender As Object, e As EventArgs) Handles cms_Tools.Opened
'		cms_Tools.ToggleRemove(SelectedRowsCount > 0)
'		cms_Tools.ToggleEdit(SelectedRowsCount > 0)
'	End Sub

'	Private Sub EditPerson() Handles cms_Tools.EditSelected
'		If SelectedRowsCount < 1 Then
'			Return
'		End If

'		For Each row As DataGridViewRow In SelectedCustomers
'			CellClicked(Me, New DataGridViewCellEventArgs(dgc_Edit.Index, row.Index))
'		Next
'	End Sub

'	Private Sub RefreshView() Handles cms_Tools.RefreshView
'		RaiseEvent RefreshDisplay()
'	End Sub
'End Class

File: M3Tools/Controls/People/Customers/CustomersDataGrid.cs
using System.ComponentModel;
using System.Windows.Forms;

namespace SPPBC.M3Tools.Data
{
	/// <summary>
	/// Custom data grid to use for displaying customer information
	/// </summary>
	public partial class CustomerDataGrid
	{
		// The columns used in this datagrid
		private readonly DataGridViewTextBoxColumn dgc_Name = new();
		private readonly DataGridViewTextBoxColumn dgc_Address = new();
		private readonly DataGridViewTextBoxColumn dgc_Phone = new();
		private readonly DataGridViewTextBoxColumn dgc_Email = new();
		private readonly DataGridViewTextBoxColumn dgc_Join = new();

		/// <summary>
		/// Event that occurs when a customer is being added to the database
		/// </summary>
		public event Events.Customers.CustomerEventHandler AddCustomer;

		/// <summary>
		/// Event that occurs when a customer is being removed from the database
		/// </summary>
		public event Events.Customers.CustomerEventHandler RemoveCustomer;

		/// <summary>
		/// Event that occurs when a customer is being updated in the database
		/// </summary>
		public event Events.Customers.CustomerEventHandler UpdateCustomer;

		/// <summary>
		/// The list of customers currently in the data grid
		/// </summary>
		[Browsable(false)]
		public Types.CustomerCollection Customers
		{
			get => Types.CustomerCollection.Cast(bsCustomers.List);
			set
			{
				if (DesignMode)
				{
					return;
				}

				bsCustomers.DataSource = value;
			}
		}

		/// <summary>
		/// The filter to place on the data in the grid
		/// </summary>
		public string Filter
		{
			get => bsCustomers.Filter;
			set => bsCustomers.Filter = value;
		}

		/// <summary>
		/// 
		/// </summary>
		[Browsable(false)]
		public Types.CustomerCollection SelectedCustomers => Types.CustomerCollection.Cast(base.SelectedRows);

		/// <summary>
		/// 
		/// </summary>
		public CustomerDataGrid() : base()
		{
			InitializeComponent();

			LoadColumns();

			AddEntry += (sender, e) => AddCustomer?.Invoke(sender, new(e.Value, e.EventType));
			UpdateEntry += (sender, e) => UpdateCustomer?.Invoke(sender, new(e.Value, e.EventType));
			RemoveEntry += (sender, e) => RemoveCustomer?.Invoke(sender, new(e.Value, e.EventType));
		}

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		protected override void LoadColumns()
		{
			base.LoadColumns();

			dgc_ID.HeaderText = "CustomerID";

			// Name Column
			dgc_Name.AutoSizeMode = System.Windows.Forms.DataGridViewAutoSizeColumnMode.DisplayedCells;
			dgc_Name.DataPropertyName = "Name";
			dgc_Name.FillWeight = 25F;
			dgc_Name.HeaderText = "Name";
			dgc_Name.MinimumWidth = 10;
			dgc_Name.Name = "dgc_Name";

			// Address Column
			dgc_Address.AutoSizeMode = System.Windows.Forms.DataGridViewAutoSizeColumnMode.Fill;
			dgc_Address.DataPropertyName = "Address";
			dgc_Address.FillWeight = 25F;
			dgc_Address.HeaderText = "Address";
			dgc_Address.MinimumWidth = 10;
			dgc_Address.Name = "dgc_Address";

			// Phone Column
			dgc_Phone.AutoSizeMode = System.Windows.Forms.DataGridViewAutoSizeColumnMode.Fill;
			dgc_Phone.DataPropertyName = "Phone";
			dgc_Phone.FillWeight = 25F;
			dgc_Phone.HeaderText = "Phone";
			dgc_Phone.MinimumWidth = 10;
			dgc_Phone.Name = "dgc_Phone";

			// Email Column
			dgc_Email.AutoSizeMode = System.Windows.Forms.DataGridViewAutoSizeColumnMode.Fill;
			dgc_Email.DataPropertyName = "Email";
			dgc_Email.FillWeight = 25F;
			dgc_Email.HeaderText = "Email";
			dgc_Email.MinimumWidth = 10;
			dgc_Email.Name = "dgc_Email";

			// Join Date Column
			dgc_Join.AutoSizeMode = System.Windows.Forms.DataGridViewAutoSizeColumnMode.AllCells;
			dgc_Join.DataPropertyName = "Joined";
			dgc_Join.DefaultCellStyle = new() { Alignment = System.Windows.Forms.DataGridViewContentAlignment.MiddleCenter, Format = "d", NullValue = "N/A", WrapMode = System.Windows.Forms.DataGridViewTriState.False };
			dgc_Join.HeaderText = "Joined";
			dgc_Join.MinimumWidth = 10;
			dgc_Join.Name = "dgc_Join";

			Columns.AddRange(
			[
				dgc_Selection, dgc_ID,
				dgc_Name, dgc_Address, dgc_Phone, dgc_Email,
				dgc_Join, dgc_Edit, dgc_Remove

			]);
		}
	}

}
